components/header.js

```diff
@@ -1,67 +1,189 @@
-const template = document.createElement("template");
-template.innerHTML = `
-  <style>
-    :host {
-      position: fixed;
-      top: 0;
-      left: 0;
-      padding: 20px 0;
-      width: 100%;
-      border-bottom: 1px solid #D3D3D3;
-      background-color: rgba(255, 255, 255, 0.8);
-    }
+import { css, html } from "../html-css-utils.js";
 
-    header {
-      width: 80%;
-      margin: 0 auto;
-      justify-content: space-between;
-      display: flex;
-      align-items: center;
-    }
+class Header extends HTMLElement {
+  constructor() {
+    super();
+    this.render();
+    this.setupEventListeners();
+    this.handleResize(); // Ensure the menu is closed initially if needed
+  }
 
-    .buttons-container {
-      display: flex;
-      align-items: center;
-      column-gap: 40px;
-    }
+  setupEventListeners() {
+    const menuButton = this.shadowRoot.querySelector('.menu-button');
+    menuButton.addEventListener('click', this.toggleMenu.bind(this));
+    window.addEventListener('resize', this.handleResize.bind(this));
+  }
 
-    a {
-      display: flex;
-      gap: 10px;
-      align-items: center;
-      text-decoration: none;
-    }
+  toggleMenu() {
+    const buttonLinks = this.shadowRoot.querySelector('.buttons-container');
+    const menuIcon = this.shadowRoot.querySelector('.menu-button img');
+    const header = this.shadowRoot.querySelector('header');
 
-    a img {
-      width: 45px;
-      height: 22.5px;
+    buttonLinks.classList.toggle('open');
+    header.classList.toggle('vertical');
+    this.classList.toggle('menu-open');
+
+    if (menuIcon.src.includes('menu.png')) {
+      menuIcon.src = 'images/cancel.png';
+    } else {
+      menuIcon.src = 'images/menu.png';
     }
+  }
```

Comment:
@SamTheKorean 샘님 제가 말씀드렸던 transition을 구현하려면 토글 될 때마다 해당 <img> 엘리먼트에 자동으로 transition이 추가되도록 아래와 같은 css 속성을 추가하면 될 것 같아요.

```css
button img {
    transition: transform 200ms ease-in-out;
}
```

그리고 옵셔널로 아이콘이 클릭될 때마다 클래스 네임을 추가하고 특정 애니메이션 효과를 추가할 수도 있습니다.

```suggestion
if (menuIcon.src.includes('menu.png')) {
      menuIcon.src = 'images/cancel.png';
      menuIcon.classList.add('menu-clicked');
    } else {
      menuIcon.src = 'images/menu.png';
      menuIcon.classList.remove('menu-clicked');
  }
```

```css
.menu-clicked {
  transform: rotate(180deg);
}
```

Reply:
실제로 돌아가는 메뉴 버튼이 클릭되면 각도가 돌아가게 하는걸 의도하신걸까요? 아니면 그냥 예시일까요?

Reply:
버튼에 ``csstransition: transform 200ms ease-in-out;```

Reply:
네 예시로 보여드린 코드입니다 :^)
Transition 속성이 적용되려면 <img> 엘리먼트의 show/hide 동작을 css에서 처리를 해줘야 하고, 그러려면 자바스크립트에서는 클래스 네임을 토글하는 방식으로 수정하는 것이 어떨까 싶습니다. 어떻게 생각하시나요?

Reply:
넵넵 말씀하신대로 구현하겠습니다!

Reply:
https://github.com/DaleStudy/leetcode-website/assets/104721736/098c04cf-9817-4018-9492-67a594eace83

의도하신게 이렇게 fade in fade out 에니메이션이 맞을까요?

Reply:
네네 정확하게 맞습니다! 수고 많으셨어요 👏👏